pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -470,262 +559,611 @@ fn storage_sync_loop<
+                    index
                         .write()
                         .await
-                        .set_awaits_download(&task.sync_id, false)
+                        .set_awaits_download(&sync_id, false)
                         .expect("timeline should be present in remote index");
-                    None
+                    register_sync_status(sync_start, task_name, Some(true));
+                    Some(TimelineSyncStatusUpdate::Downloaded)
                 }
-                DownloadedTimeline::FailedAndRescheduled => {
-                    register_sync_status(sync_start, sync_name, Some(false));
+                Err(e) => {
+                    error!("Failed to update local timeline metadata: {:?}", e);
+                    download_data.retries += 1;
+                    sync_queue::push(sync_id, SyncTask::Download(download_data));
+                    register_sync_status(sync_start, task_name, Some(false));
                     None
                 }
-                DownloadedTimeline::Successful => {
-                    register_sync_status(sync_start, sync_name, Some(true));
-                    remote_index
-                        .write()
-                        .await
-                        .set_awaits_download(&task.sync_id, false)
-                        .expect("timeline should be present in remote index");
-                    Some(TimelineSyncStatusUpdate::Downloaded)
+            }
+        }
+    }
+}
+
+async fn update_local_metadata(
+    conf: &'static PageServerConf,
+    sync_id: ZTenantTimelineId,
+    remote_timeline: Option<&RemoteTimeline>,
+) -> anyhow::Result<()> {
+    let remote_metadata = match remote_timeline {
+        Some(timeline) => &timeline.metadata,
+        None => {
+            info!("No remote timeline to update local metadata from, skipping the update");
+            return Ok(());
+        }
+    };
+    let remote_lsn = remote_metadata.disk_consistent_lsn();
+
+    let local_metadata_path = metadata_path(conf, sync_id.timeline_id, sync_id.tenant_id);
+    let local_lsn = if local_metadata_path.exists() {
+        let local_metadata = read_metadata_file(&local_metadata_path)
+            .await
+            .with_context(|| {
+                format!(
+                    "Failed to load local metadata from path '{}'",
+                    local_metadata_path.display()
+                )
+            })?;
+
+        Some(local_metadata.disk_consistent_lsn())
+    } else {
+        None
+    };
+
+    if local_lsn < Some(remote_lsn) {
+        info!(
+            "Updating local timeline metadata from remote timeline: local disk_consistent_lsn={:?}, remote disk_consistent_lsn={}",
+            local_lsn, remote_lsn
+        );
+
+        let remote_metadata_bytes = remote_metadata
+            .to_bytes()
+            .context("Failed to serialize remote metadata to bytes")?;
+        fs::write(&local_metadata_path, &remote_metadata_bytes)
```

Comment:
I wonder do we need to duplicate this fsync dance from here https://github.com/neondatabase/neon/blob/main/pageserver/src/layered_repository.rs#L576

Reply:
We do not fsync downloaded layers either. I wonder what is the right thing to do. There are several options. Option one is to download layer files and perform a durable rename (after fsync), if during startup pageserver sees partial files they can be safely removed. Second option is to avoid rename but detect "future" layer files by parsing a file name and comparing the lsn with locally available metadata. This can be done separately, I can implement that on top of this PR. Thoughts?

Reply:
Yes, we do need `fsync`. I kept this as a part of the checksums task, since very related IMO, but we can create a separate issue for that.

Either way, not excited to add `fsync`s right now.